### ISSUE TITLE
[BUG] Submitting a Spark Application Fails on K8s versions < 1.21

### ISSUE THREAD

**Issue by Amri91** [kubeflow/spark-operator#2051]
## Description

According to the docs, versions 1.16+ should be supported. However, I am unable to make submitting a spark application work on K8s 1.17, 1.18, 1.19, and 1.20. It works on 1.21, but unfortunately, this is not an option for me.

- [x] ✋ I have searched the open/closed issues and my issue is not listed.

## Reproduction Code [Required]

1. minikube start --kubernetes-version=v1.17.0 --cpus=10 --memory=6000
1. git clone https://github.com/kubeflow/spark-operator.git && cd spark-operator
1. helm install spark-operator ./charts/spark-operator-chart --namespace spark-operator --create-namespace
1. minikube kubectl -- apply -f ./examples/spark-pi.yaml (I updated the namespace to spark-operator and double check the service account is spark-operator-spark)

## Expected behavior

The spark application deploys successfully.

## Actual behavior

State: SUBMISSION_FAILED

### Terminal Output Screenshot(s)

```
{
"lastSubmissionAttemptTime": "2024-06-07T17:48:55Z",
"terminationTime": null,
"driverInfo": {},
"applicationState": {
"state": "SUBMISSION_FAILED",
"errorMessage": "failed to run spark-submit for SparkApplication spark-operator/spark-pi: 24/06/07 17:48:53 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n24/06/07 17:48:53 INFO SparkKubernetesClientFactory: Auto-configuring K8S client using current context from users K8S config file\n24/06/07 17:48:53 INFO KerberosConfDriverFeatureStep: You have not specified a krb5.conf file locally or via a ConfigMap. Make sure that you have the krb5.conf locally on the driver image.\nException in thread \"main\" io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://10.96.0.1:443/api/v1/namespaces/spark-operator/services/spark-pi-c268088ff3d15554-driver-svc?fieldManager=fabric8\u0026force=true. Message: failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\" (and doesn't have default value). Received status: Status(apiVersion=v1, code=500, details=null, kind=Status, message=failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\" (and doesn't have default value), metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=null, status=Failure, additionalProperties={}).\n\tat io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:518)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:535)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:430)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:408)\n\tat io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handlePatch(BaseOperation.java:713)\n\tat io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:232)\n\tat io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:237)\n\tat io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:252)\n\tat io.fabric8.kubernetes.client.dsl.internal.BaseOperation.serverSideApply(BaseOperation.java:1132)\n\tat io.fabric8.kubernetes.client.dsl.internal.BaseOperation.serverSideApply(BaseOperation.java:92)\n\tat io.fabric8.kubernetes.client.extension.ResourceAdapter.serverSideApply(ResourceAdapter.java:325)\n\tat java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)\n\tat java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)\n\tat java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)\n\tat java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)\n\tat java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)\n\tat java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)\n\tat java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)\n\tat io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.performOperation(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:294)\n\tat io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.serverSideApply(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:324)\n\tat io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.serverSideApply(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:55)\n\tat org.apache.spark.deploy.k8s.submit.Client.run(KubernetesClientApplication.scala:176)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$6(KubernetesClientApplication.scala:256)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$6$adapted(KubernetesClientApplication.scala:250)\n\tat org.apache.spark.util.SparkErrorUtils.tryWithResource(SparkErrorUtils.scala:48)\n\tat org.apache.spark.util.SparkErrorUtils.tryWithResource$(SparkErrorUtils.scala:46)\n\tat org.apache.spark.util.Utils$.tryWithResource(Utils.scala:94)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.run(KubernetesClientApplication.scala:250)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.start(KubernetesClientApplication.scala:223)\n\tat org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:1029)\n\tat org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:194)\n\tat org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:217)\n\tat org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:91)\n\tat org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1120)\n\tat org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1129)\n\tat org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)\nCaused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://10.96.0.1:443/api/v1/namespaces/spark-operator/services/spark-pi-c268088ff3d15554-driver-svc?fieldManager=fabric8\u0026force=true. Message: failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\" (and doesn't have default value). Received status: Status(apiVersion=v1, code=500, details=null, kind=Status, message=failed to create typed patch object: .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\" (and doesn't have default value), metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=null, status=Failure, additionalProperties={}).\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:671)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:651)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:600)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:560)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:140)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:52)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:137)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\n24/06/07 17:48:55 INFO ShutdownHookManager: Shutdown hook called\n24/06/07 17:48:55 INFO ShutdownHookManager: Deleting directory /tmp/spark-55d9835e-0694-477b-bbd3-6e96ee5af368\n"
}
```

## Environment & Versions

- Spark Operator App version: v1beta2-1.6.0-3.5.0
- Helm Chart Version: 1.4.0
- Kubernetes Version: 1.17 - 1.20
- Apache Spark version: 3.5.1

## Additional context

Other attempts

Attempt 1:

1. move to branch v1beta2-1.3.3-3.1.1
1. Update the dockerfile SPARK_IMAGE argument to spark:3.5.1 since the existing URL did not work.
1. Built the image
1. Loaded the image using minikube load

I get the same error.

Attempt 2:

1. moved to branch v1beta2-1.1.2-2.4.5
1. Updated the docker image to use root and spark 3.5.1, built it, and loaded it.
1. cloned https://github.com/helm/charts.git and deployed it. I had to tweak the cluster role a bit to make it work.

Same error:

```
{
  "lastSubmissionAttemptTime": "2024-06-07T20:22:20Z",
  "terminationTime": null,
  "driverInfo": {},
  "applicationState": {
    "state": "SUBMISSION_FAILED",
    "errorMessage": "failed to run spark-submit for SparkApplication default/spark-pi-v3: 24/06/07 20:22:19 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n24/06/07 20:22:19 INFO SparkKubernetesClientFactory: Auto-configuring K8S client using current context from users K8S config file\n24/06/07 20:22:19 INFO KerberosConfDriverFeatureStep: You have not specified a krb5.conf file locally or via a ConfigMap. Make sure that you have the krb5.conf locally on the driver image.\nException in thread \"main\" io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://10.96.0.1:443/api/v1/namespaces/default/services/spark-pi-v3-897dad8ff45dccf8-driver-svc?fieldManager=fabric8\u0026force=true. Message: failed to create typed patch object: errors:\n  .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 1: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 2: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ipFamilies: schema error: invalid atom: inlined\n  .spec.ipFamilyPolicy: schema error: invalid atom: inlined. Received status: Status(apiVersion=v1, code=500, details=null, kind=Status, message=failed to create typed patch object: errors:\n  .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 1: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 2: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ipFamilies: schema error: invalid atom: inlined\n  .spec.ipFamilyPolicy: schema error: invalid atom: inlined, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=null, status=Failure, additionalProperties={}).\n\tat io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:518)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:535)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:430)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:408)\n\tat io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handlePatch(BaseOperation.java:713)\n\tat io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:232)\n\tat io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:237)\n\tat io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:252)\n\tat io.fabric8.kubernetes.client.dsl.internal.BaseOperation.serverSideApply(BaseOperation.java:1132)\n\tat io.fabric8.kubernetes.client.dsl.internal.BaseOperation.serverSideApply(BaseOperation.java:92)\n\tat io.fabric8.kubernetes.client.extension.ResourceAdapter.serverSideApply(ResourceAdapter.java:325)\n\tat java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)\n\tat java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)\n\tat java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)\n\tat java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)\n\tat java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)\n\tat java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)\n\tat java.base/java.util.stream.ReferencePipeline.collect(Unknown Source)\n\tat io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.performOperation(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:294)\n\tat io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.serverSideApply(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:324)\n\tat io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.serverSideApply(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:55)\n\tat org.apache.spark.deploy.k8s.submit.Client.run(KubernetesClientApplication.scala:176)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$6(KubernetesClientApplication.scala:256)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$6$adapted(KubernetesClientApplication.scala:250)\n\tat org.apache.spark.util.SparkErrorUtils.tryWithResource(SparkErrorUtils.scala:48)\n\tat org.apache.spark.util.SparkErrorUtils.tryWithResource$(SparkErrorUtils.scala:46)\n\tat org.apache.spark.util.Utils$.tryWithResource(Utils.scala:94)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.run(KubernetesClientApplication.scala:250)\n\tat org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.start(KubernetesClientApplication.scala:223)\n\tat org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:1029)\n\tat org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:194)\n\tat org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:217)\n\tat org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:91)\n\tat org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1120)\n\tat org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1129)\n\tat org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)\nCaused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://10.96.0.1:443/api/v1/namespaces/default/services/spark-pi-v3-897dad8ff45dccf8-driver-svc?fieldManager=fabric8\u0026force=true. Message: failed to create typed patch object: errors:\n  .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 1: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 2: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ipFamilies: schema error: invalid atom: inlined\n  .spec.ipFamilyPolicy: schema error: invalid atom: inlined. Received status: Status(apiVersion=v1, code=500, details=null, kind=Status, message=failed to create typed patch object: errors:\n  .spec.ports: element 0: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 1: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ports: element 2: associative list with keys has an element that omits key field \"protocol\"\n  .spec.ipFamilies: schema error: invalid atom: inlined\n  .spec.ipFamilyPolicy: schema error: invalid atom: inlined, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=null, status=Failure, additionalProperties={}).\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:671)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:651)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:600)\n\tat io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:560)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:140)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:52)\n\tat java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)\n\tat java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)\n\tat io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:137)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)\n\tat java.base/java.lang.Thread.run(Unknown Source)\n24/06/07 20:22:20 INFO ShutdownHookManager: Shutdown hook called\n24/06/07 20:22:20 INFO ShutdownHookManager: Deleting directory /tmp/spark-38ae3f3f-f4c7-419f-b30a-6f629dc23239\n"
  },
  "submissionAttempts": 1
}
```

**Comment by Amri91** [comment]
I resolved this by using Spark 3.4.1

**Comment by zhang5059T** [comment]
Hi @Amri91 I use v1beta2-1.3.3-3.1.1 rebuild it and encountered another problem. When I changed the usage of the path of Example, a strange exception occurred.
## Exception
24/06/14 02:45:56 ERROR CoarseGrainedExecutorBackend: Executor self-exiting due to : Unable to create executor due to ./spark-examples_2.12-3.3.0.jar.jar
java.nio.file.AccessDeniedException: ./spark-examples_2.12-3.3.0.jarjar

Have you come across this problem?
